### ISSUE TITLE
timeout fetching nefiteasy data with version 0.4.2

### ISSUE THREAD

**Issue by mchangsp** [ksya/ha-nefiteasy#298]
Every now and then the nefiteasy integration is offline.
This is the log with Home Assistant 2024.1.5 and nefiteasy 0.4.2

```
2024-01-31 09:57:47.006 DEBUG (MainThread) [custom_components.nefiteasy] Finished fetching nefiteasy data in 5.041 seconds (success: True)
2024-01-31 09:58:51.496 DEBUG (MainThread) [custom_components.nefiteasy] Finished fetching nefiteasy data in 4.532 seconds (success: True)
2024-01-31 09:59:59.966 ERROR (MainThread) [custom_components.nefiteasy] Timeout fetching nefiteasy data
2024-01-31 09:59:59.966 DEBUG (MainThread) [custom_components.nefiteasy] Finished fetching nefiteasy data in 9.002 seconds (success: False)
2024-01-31 10:00:30.035 DEBUG (MainThread) [custom_components.nefiteasy] Shutdown connection to Bosch cloud
2024-01-31 10:00:30.035 DEBUG (MainThread) [custom_components.nefiteasy] Initialize Nefit class
2024-01-31 10:00:30.062 DEBUG (MainThread) [custom_components.nefiteasy] Start connecting.
2024-01-31 10:00:30.062 DEBUG (MainThread) [custom_components.nefiteasy] Waiting for connected event.
2024-01-31 10:00:30.369 DEBUG (MainThread) [custom_components.nefiteasy] Connected successfully.
2024-01-31 10:00:59.371 DEBUG (MainThread) [custom_components.nefiteasy] Did not get a response in time for testing connection.
2024-01-31 10:00:59.371 DEBUG (MainThread) [custom_components.nefiteasy] Successfully verified connection.
2024-01-31 10:01:04.866 DEBUG (MainThread) [custom_components.nefiteasy] Initialize Nefit class
2024-01-31 10:01:04.904 DEBUG (MainThread) [custom_components.nefiteasy] Start connecting.
2024-01-31 10:01:04.905 DEBUG (MainThread) [custom_components.nefiteasy] Waiting for connected event.
2024-01-31 10:01:05.239 DEBUG (MainThread) [custom_components.nefiteasy] Unexpected disconnect of 308010606 with Bosch server. Try to reconnect..
2024-01-31 10:01:05.243 DEBUG (MainThread) [custom_components.nefiteasy] Connected successfully.
2024-01-31 10:01:34.246 DEBUG (MainThread) [custom_components.nefiteasy] Did not get a response in time for testing connection.
2024-01-31 10:01:34.246 DEBUG (MainThread) [custom_components.nefiteasy] Successfully verified connection.
2024-01-31 10:01:44.557 DEBUG (MainThread) [custom_components.nefiteasy] Initialize Nefit class
2024-01-31 10:01:44.593 DEBUG (MainThread) [custom_components.nefiteasy] Start connecting.
2024-01-31 10:01:44.593 DEBUG (MainThread) [custom_components.nefiteasy] Waiting for connected event.
2024-01-31 10:01:44.930 DEBUG (MainThread) [custom_components.nefiteasy] Unexpected disconnect of 308010606 with Bosch server. Try to reconnect..
2024-01-31 10:01:44.937 DEBUG (MainThread) [custom_components.nefiteasy] Connected successfully.
2024-01-31 10:02:13.941 DEBUG (MainThread) [custom_components.nefiteasy] Did not get a response in time for testing connection.
2024-01-31 10:02:13.942 DEBUG (MainThread) [custom_components.nefiteasy] Successfully verified connection.
2024-01-31 10:02:34.263 DEBUG (MainThread) [custom_components.nefiteasy] Initialize Nefit class
2024-01-31 10:02:34.305 DEBUG (MainThread) [custom_components.nefiteasy] Start connecting.
2024-01-31 10:02:34.306 DEBUG (MainThread) [custom_components.nefiteasy] Waiting for connected event.
2024-01-31 10:02:34.597 DEBUG (MainThread) [custom_components.nefiteasy] Unexpected disconnect of 308010606 with Bosch server. Try to reconnect..
2024-01-31 10:02:34.604 DEBUG (MainThread) [custom_components.nefiteasy] Connected successfully.
2024-01-31 10:03:03.607 DEBUG (MainThread) [custom_components.nefiteasy] Did not get a response in time for testing connection.
2024-01-31 10:03:03.607 DEBUG (MainThread) [custom_components.nefiteasy] Successfully verified connection.
2024-01-31 10:03:43.725 DEBUG (MainThread) [custom_components.nefiteasy] Initialize Nefit class
2024-01-31 10:03:43.768 DEBUG (MainThread) [custom_components.nefiteasy] Start connecting.
2024-01-31 10:03:43.768 DEBUG (MainThread) [custom_components.nefiteasy] Waiting for connected event.
2024-01-31 10:03:44.052 DEBUG (MainThread) [custom_components.nefiteasy] Unexpected disconnect of 308010606 with Bosch server. Try to reconnect..
2024-01-31 10:03:44.055 DEBUG (MainThread) [custom_components.nefiteasy] Connected successfully.
2024-01-31 10:04:13.058 DEBUG (MainThread) [custom_components.nefiteasy] Did not get a response in time for testing connection.
2024-01-31 10:04:13.058 DEBUG (MainThread) [custom_components.nefiteasy] Successfully verified connection.
2024-01-31 10:05:33.294 DEBUG (MainThread) [custom_components.nefiteasy] Initialize Nefit class
2024-01-31 10:05:33.320 DEBUG (MainThread) [custom_components.nefiteasy] Start connecting.
2024-01-31 10:05:33.321 DEBUG (MainThread) [custom_components.nefiteasy] Waiting for connected event.
2024-01-31 10:05:33.622 DEBUG (MainThread) [custom_components.nefiteasy] Unexpected disconnect of 308010606 with Bosch server. Try to reconnect..
2024-01-31 10:05:33.626 DEBUG (MainThread) [custom_components.nefiteasy] Connected successfully.
2024-01-31 10:06:02.629 DEBUG (MainThread) [custom_components.nefiteasy] Did not get a response in time for testing connection.
2024-01-31 10:06:02.629 DEBUG (MainThread) [custom_components.nefiteasy] Successfully verified connection.
2024-01-31 10:07:22.751 DEBUG (MainThread) [custom_components.nefiteasy] Initialize Nefit class
2024-01-31 10:07:22.789 DEBUG (MainThread) [custom_components.nefiteasy] Start connecting.
2024-01-31 10:07:22.789 DEBUG (MainThread) [custom_components.nefiteasy] Waiting for connected event.
2024-01-31 10:07:23.061 DEBUG (MainThread) [custom_components.nefiteasy] Unexpected disconnect of 308010606 with Bosch server. Try to reconnect..
2024-01-31 10:07:23.065 DEBUG (MainThread) [custom_components.nefiteasy] Connected successfully.
2024-01-31 10:07:23.771 DEBUG (MainThread) [custom_components.nefiteasy] Finished fetching nefiteasy data in 0.426 seconds (success: True)
2024-01-31 10:07:26.712 DEBUG (MainThread) [custom_components.nefiteasy] Finished fetching nefiteasy data in 2.936 seconds (success: True)
2024-01-31 10:07:26.729 WARNING (MainThread) [homeassistant.helpers.entity] Entity None (<class 'custom_components.nefiteasy.climate.NefitThermostat'>) is using deprecated supported features values which will be removed in HA Core 2025.1. Instead it should use <ClimateEntityFeature.TARGET_TEMPERATURE|PRESET_MODE: 17>, please create a bug report at https://github.com/ksya/ha-nefiteasy/issues and reference https://developers.home-assistant.io/blog/2023/12/28/support-feature-magic-numbers-deprecation
2024-01-31 10:07:41.501 DEBUG (MainThread) [custom_components.nefiteasy] Finished fetching nefiteasy data in 4.770 seconds (success: True)
2024-01-31 10:08:45.889 DEBUG (MainThread) [custom_components.nefiteasy] Finished fetching nefiteasy data in 4.763 seconds (success: True)
2024-01-31 10:09:51.159 DEBUG (MainThread) [custom_components.nefiteasy] Finished fetching nefiteasy data in 5.034 seconds (success: True)
2024-01-31 10:10:55.921 DEBUG (MainThread) [custom_components.nefiteasy] Finished fetching nefiteasy data in 4.795 seconds (success: True)
```
ENDQYUOTE

**Comment by RobBie1221** [comment]
It makes no sense to me...
Once it sends out the "get brand ID" command for verifying connection, there is a timeout of 29 seconds before a timeout is given after which the connection procedure is aborted. Somehow it hangs, but I can't see why yet.

I released 0.4.3 with additional debug logging. See if you can catch the error with this newer version.

**Comment by mchangsp** [comment]
Thanks for looking into it. Will install 0.4.3 and try to catch the error

**Comment by mchangsp** [comment]
I enabled debug loggging and succeeded to catch this:

```
2024-02-01 06:57:36.410 DEBUG (MainThread) [custom_components.nefiteasy] Finished fetching nefiteasy data in 3.297 seconds (success: True)
2024-02-01 06:58:39.420 DEBUG (MainThread) [custom_components.nefiteasy] Finished fetching nefiteasy data in 3.308 seconds (success: True)
2024-02-01 06:59:43.175 DEBUG (MainThread) [custom_components.nefiteasy] Finished fetching nefiteasy data in 4.063 seconds (success: True)
2024-02-01 07:00:52.115 ERROR (MainThread) [custom_components.nefiteasy] Timeout fetching nefiteasy data
2024-02-01 07:00:52.115 DEBUG (MainThread) [custom_components.nefiteasy] Finished fetching nefiteasy data in 9.003 seconds (success: False)
2024-02-01 07:01:22.178 DEBUG (MainThread) [custom_components.nefiteasy] Shutdown connection to Bosch cloud
2024-02-01 07:01:22.178 DEBUG (MainThread) [custom_components.nefiteasy] Initialize Nefit class
2024-02-01 07:01:22.213 DEBUG (MainThread) [custom_components.nefiteasy] Start connecting.
2024-02-01 07:01:22.213 DEBUG (MainThread) [custom_components.nefiteasy] Set is connecting to true
2024-02-01 07:01:22.213 DEBUG (MainThread) [custom_components.nefiteasy] Waiting for connected event.
2024-02-01 07:01:22.598 DEBUG (MainThread) [custom_components.nefiteasy] Connected successfully.
2024-02-01 07:01:22.599 DEBUG (MainThread) [custom_components.nefiteasy] Wait for message event
2024-02-01 07:01:39.867 ERROR (SyncWorker_20) [homeassistant] Error doing job: Task was destroyed but it is pending!
2024-02-01 07:01:51.600 DEBUG (MainThread) [custom_components.nefiteasy] Did not get a response in time for testing connection.
2024-02-01 07:01:51.600 DEBUG (MainThread) [custom_components.nefiteasy] Successfully verified connection.
2024-02-01 07:01:51.600 DEBUG (MainThread) [custom_components.nefiteasy] Set is connecting to false
2024-02-01 07:01:56.904 DEBUG (MainThread) [custom_components.nefiteasy] Initialize Nefit class
2024-02-01 07:01:56.938 DEBUG (MainThread) [custom_components.nefiteasy] Start connecting.
2024-02-01 07:01:56.939 DEBUG (MainThread) [custom_components.nefiteasy] Set is connecting to true
2024-02-01 07:01:56.939 DEBUG (MainThread) [custom_components.nefiteasy] Waiting for connected event.
2024-02-01 07:01:57.214 DEBUG (MainThread) [custom_components.nefiteasy] Connected successfully.
2024-02-01 07:01:57.215 DEBUG (MainThread) [custom_components.nefiteasy] Wait for message event
2024-02-01 07:02:26.217 DEBUG (MainThread) [custom_components.nefiteasy] Did not get a response in time for testing connection.
2024-02-01 07:02:26.217 DEBUG (MainThread) [custom_components.nefiteasy] Successfully verified connection.
2024-02-01 07:02:26.218 DEBUG (MainThread) [custom_components.nefiteasy] Set is connecting to false
2024-02-01 07:02:36.624 DEBUG (MainThread) [custom_components.nefiteasy] Initialize Nefit class
2024-02-01 07:02:36.652 DEBUG (MainThread) [custom_components.nefiteasy] Start connecting.
2024-02-01 07:02:36.652 DEBUG (MainThread) [custom_components.nefiteasy] Set is connecting to true
2024-02-01 07:02:36.652 DEBUG (MainThread) [custom_components.nefiteasy] Waiting for connected event.
2024-02-01 07:02:37.337 DEBUG (MainThread) [custom_components.nefiteasy] Connected successfully.
2024-02-01 07:02:37.339 DEBUG (MainThread) [custom_components.nefiteasy] Wait for message event
2024-02-01 07:03:06.341 DEBUG (MainThread) [custom_components.nefiteasy] Did not get a response in time for testing connection.
2024-02-01 07:03:06.341 DEBUG (MainThread) [custom_components.nefiteasy] Successfully verified connection.
2024-02-01 07:03:06.341 DEBUG (MainThread) [custom_components.nefiteasy] Set is connecting to false
2024-02-01 07:03:26.524 DEBUG (MainThread) [custom_components.nefiteasy] Initialize Nefit class
2024-02-01 07:03:26.564 DEBUG (MainThread) [custom_components.nefiteasy] Start connecting.
2024-02-01 07:03:26.564 DEBUG (MainThread) [custom_components.nefiteasy] Set is connecting to true
2024-02-01 07:03:26.564 DEBUG (MainThread) [custom_components.nefiteasy] Waiting for connected event.
2024-02-01 07:03:26.853 DEBUG (MainThread) [custom_components.nefiteasy] Connected successfully.
2024-02-01 07:03:26.854 DEBUG (MainThread) [custom_components.nefiteasy] Wait for message event
2024-02-01 07:03:27.046 DEBUG (MainThread) [custom_components.nefiteasy] Message event received
2024-02-01 07:03:27.046 DEBUG (MainThread) [custom_components.nefiteasy] Set is connecting to false
2024-02-01 07:03:27.351 DEBUG (MainThread) [custom_components.nefiteasy] Finished fetching nefiteasy data in 0.305 seconds (success: True)
2024-02-01 07:03:29.550 DEBUG (MainThread) [custom_components.nefiteasy] Finished fetching nefiteasy data in 2.191 seconds (success: True)
2024-02-01 07:03:29.578 WARNING (MainThread) [homeassistant.helpers.entity] Entity None (<class 'custom_components.nefiteasy.climate.NefitThermostat'>) is using deprecated supported features values which will be removed in HA Core 2025.1. Instead it should use <ClimateEntityFeature.TARGET_TEMPERATURE|PRESET_MODE: 17>, please create a bug report at https://github.com/ksya/ha-nefiteasy/issues and reference https://developers.home-assistant.io/blog/2023/12/28/support-feature-magic-numbers-deprecation
2024-02-01 07:03:43.000 DEBUG (MainThread) [custom_components.nefiteasy] Finished fetching nefiteasy data in 3.421 seconds (success: True)
2024-02-01 07:04:47.002 DEBUG (MainThread) [custom_components.nefiteasy] Finished fetching nefiteasy data in 3.639 seconds (success: True)
```

**Comment by mchangsp** [comment]
Please note I have the following automation running, when the logging above was done.
Without this automation the Nefit client stays disconnected.


```
alias: Nefit auto reload when unavailable
description: ""
trigger:
  - platform: state
    entity_id:
      - climate.nefit
    to: unavailable
    for:
      hours: 0
      minutes: 0
      seconds: 30
condition: []
  - service: logbook.log
    metadata: {}
    data:
      name: Nefit
      message: Nefit became unavailable, reloading
      entity_id: climate.nefit
      domain: climate
  - service: homeassistant.reload_config_entry
    target:
      entity_id: climate.nefit
    data: {}
mode: queued
max: 10
```

**Comment by mchangsp** [comment]
> It makes no sense to me... Once it sends out the "get brand ID" command for verifying connection, there is a timeout of 29 seconds before a timeout is given after which the connection procedure is aborted. Somehow it hangs, but I can't see why yet.
> 
> I released 0.4.3 with additional debug logging. See if you can catch the error with this newer version.

Oops, now I understand the timeout. I will disable my automation and see how everything goes

**Comment by AdeZwart** [comment]
I'm also still running the automation that someone suggested on one of the older issues, except I'm no longer automatically reloading the integration and only use it to notify me that it became unavailable. I don't have any hard numbers, but I guess 3 out of 4 times that I get the notification the integration does seem to recover and reconnect successfully. 

I'll also install the 0.4.3 version today and see if we can catch something 😁 

Thanks for all the efforts so far 🙌🏻

**Comment by RobBie1221** [comment]
There is an issue with the Nefit server, that’s for sure. They also don’t care because the app always reconnects. Only HA keeps the connection alive for so long and they don’t officially support it.

Having said that, the integration should always be able to reconnect once the server is responding again. Manual resets or automations should not be necessary. Question with these automations is whether they are sufficiently tolerant towards the Nefit server, so not restarting the integration automatically but giving it a few minutes is a good idea to see if it can somewhere recover.

**Comment by sanderlv** [comment]
> > It makes no sense to me... Once it sends out the "get brand ID" command for verifying connection, there is a timeout of 29 seconds before a timeout is given after which the connection procedure is aborted. Somehow it hangs, but I can't see why yet.
> > I released 0.4.3 with additional debug logging. See if you can catch the error with this newer version.
> 
> Oops, now I understand the timeout. I will disable my automation and see how everything goes

This helped. I had the automation also running but it seems now it is captured in 0.4.2 (and 3)? Now disabled the reload automation and see how it goes...
